### PR TITLE
Limit scan result to specified scopes

### LINKED ISSUE #90: scanner: Limit scans to specific dependency scopes
Add a parameter to the scanner module to select which scopes will be scanned.

### PR DESCRIPTION
**Pull request by pnowakowski**
Resolves #90

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/239)

### CODE REVIEW

File: scanner/src/main/kotlin/Main.kt [hunk 1]
```diff
@@ -99,6 +99,12 @@ object Main {
             order = PARAMETER_ORDER_OPTIONAL)
     private var inputPath: File? = null
 
+    @Parameter(description = "The list of scopes that shall be scanned for licenses. Works only with " +
```
Comment (mnonnenmacher) [inline scanner/src/main/kotlin/Main.kt]: ...with the...

File: scanner/src/main/kotlin/Main.kt [hunk 2]
```diff
@@ -213,7 +219,17 @@ object Main {
...
 
             val packages = mutableListOf(analyzerResult.project.toPackage())
-            packages.addAll(analyzerResult.packages)
+            if (includeScopes.isNotEmpty()) {
+                println("Limiting scan to scopes: $includeScopes")
+                val includedScopes = analyzerResult.project.scopes.filter { includeScopes.contains(it.name) }
+                packages.addAll(analyzerResult.packages.filter { pkg ->
+                    includedScopes.any {
+                        it.contains(pkg)
+                    }
```
Comment (mnonnenmacher) [inline scanner/src/main/kotlin/Main.kt]: Would fit into one line.

**Comment by mnonnenmacher** [comment]
Please also add a "scannedScopes: List<String>" field to the ScanSummary class and fill it with all scanned scopes.

**Comment by pnowakowski** [comment]
>  the
Should that field be filled only when `--scopes` parameter was used, and we actually limit scanner results? Or if no value for the param was passed then all  project's scopes should be used?

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[scanner/src/main/kotlin/Main.kt, line 102 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/239#-L3Xo-wq-wgKyx3QfYb8-r1-102:-L40xYt91AvJn6rPKnOU:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/b9ec12331963528304620fd782a8ed3a2e5c4dd7/scanner/src/main/kotlin/Main.kt#L102)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

...with the...
</blockquote></details>

Done.

---

*[scanner/src/main/kotlin/Main.kt, line 104 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/239#-L3Xo-wq-wgKyx3QfYb8-r1-104:-L40xZ_4v0qzu2En8xBS:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/b9ec12331963528304620fd782a8ed3a2e5c4dd7/scanner/src/main/kotlin/Main.kt#L104)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

Please call this "--scopes" and remove the short name.
</blockquote></details>

Done.

---

*[scanner/src/main/kotlin/Main.kt, line 228 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/239#-L3Xo-wq-wgKyx3QfYb8-r1-228:-L40xe7Uk2NX_9HWIbHQ:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/b9ec12331963528304620fd782a8ed3a2e5c4dd7/scanner/src/main/kotlin/Main.kt#L228)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

Would fit into one line.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/239#-:-L40yEPXnn73BltnMZ9h:b-39kohd)*

**Comment by mnonnenmacher** [comment]
It should always be filled. I just realized that we also need to track the ignored scopes, so please also add a field `ignoredScopes`.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/239#-:-L41-kZXwsKhpmSHOHoO:b-c3lf7o)*

File: scanner/src/main/kotlin/Main.kt [hunk 3]
```diff
@@ -99,6 +99,12 @@ object Main {
             order = PARAMETER_ORDER_OPTIONAL)
     private var inputPath: File? = null
 
+    @Parameter(description = "The list of scopes that shall be scanned for licenses. Works only with the " +
```
Comment (mnonnenmacher) [inline scanner/src/main/kotlin/Main.kt]: Please remove "for licenses", we could also integrate scanners that scan for something else.

**Comment by pnowakowski** [comment]
Done.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[scanner/src/main/kotlin/Main.kt, line 102 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/239#-L3Xo-wq-wgKyx3QfYb8-r2-102:-L45NBCwM-C9wRqxP28s:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/d412b298fd52bf25c0b9c16b8414019720a636b1/scanner/src/main/kotlin/Main.kt#L102)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

Please remove "for licenses", we could also integrate scanners that scan for something else.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/239#-:-L45NCAbUB9v-msPIgLR:b-896fix)*

**Comment by mnonnenmacher** [comment]
Reviewed 1 of 1 files at r2.
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[scanner/src/main/kotlin/Main.kt, line 211 at r3](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/239#-L46A_VSMKv8he5UA3Ct:-L46A_VSMKv8he5UA3Cu:b-dev6r6) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/d498e24b3e9dbcbb2cebc8f0f4cc9370fc814a5c/scanner/src/main/kotlin/Main.kt#L211)):*
> ```Kotlin
> 
>         val scannedScopesNames: MutableSet<String> = mutableSetOf()
>         val ignoredScopes: MutableSet<String> = mutableSetOf()
> ```

For constistency, please either put "...Names" to both names or to none.

---

*[scanner/src/main/kotlin/Main.kt, line 230 at r3](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/239#-L46AQ6RkBPblnz9e0XY:-L46AQ6S7uA9hxwn0wcj:b-z97a6c) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/d498e24b3e9dbcbb2cebc8f0f4cc9370fc814a5c/scanner/src/main/kotlin/Main.kt#L230)):*
> ```Kotlin
>             if (includeScopes.isNotEmpty()) {
>                 println("Limiting scan to scopes: $includeScopes")
>                 val includedScopesMap = analyzerResult.project.scopes.groupBy { includeScopes.contains(it.name) }
> ```

If you do `...scopes.map { it.name }.groupBy...`here you can simplify the code below. But maybe it wouldn't work, see my next comment.

---

*[scanner/src/main/kotlin/Main.kt, line 236 at r3](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/239#-L46BI_YPS3xs5rxBOUr:-L46BI_ZfeMPd7nW4e0Z:b-mf72mc) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/d498e24b3e9dbcbb2cebc8f0f4cc9370fc814a5c/scanner/src/main/kotlin/Main.kt#L236)):*
> ```Kotlin
>                     scannedScopesNames.addAll(includedScopesNames.map { it.name })
>                     packages.addAll(
>                             analyzerResult.packages.filter { pkg -> includedScopesNames.any { it.contains(pkg) } })
> ```

This would only work if `includedScopesNames`would contain `Scope`objects instead of Strings, right?

---

*[scanner/src/main/kotlin/Main.kt, line 238 at r3](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/239#-L46BTazX1o-xW7U0CyI:-L46BTazX1o-xW7U0CyJ:bdi67o) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/d498e24b3e9dbcbb2cebc8f0f4cc9370fc814a5c/scanner/src/main/kotlin/Main.kt#L238)):*
> ```Kotlin
>                             analyzerResult.packages.filter { pkg -> includedScopesNames.any { it.contains(pkg) } })
>                 } else {
>                     println("No scopes found for given names, scanning whole project.")
> ```

I wouldn't do this fallback, it's well-intentioned, but in the end it might be desired to scan nothing if a project does not contain a specific scope.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/239)*

**Comment by mnonnenmacher** [comment]
Reviewed 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/239)*

**Pull request merged**